File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

public class Solution {

    // 주식 가격이 떨어지지 않은 기간을 계산하는 메서드
    public static int[] solution(int[] prices) {
        int n = prices.length; // 주식 가격 배열의 길이
        int[] answer = new int[n]; // 결과 배열 초기화

        // 가격이 떨어질 가능성이 있는 인덱스를 저장할 스택
        Deque<Integer> stack = new ArrayDeque<>();

        // 주식 가격 배열을 순회하며 가격이 떨어지는 순간을 찾음
        for (int i = 0; i < n; i++) {
            // 스택이 비어있지 않고, 현재 가격이 스택 최상단 가격보다 낮으면
            while (!stack.isEmpty() && prices[stack.peek()] > prices[i]) {
                int j = stack.pop(); // 스택에서 인덱스 꺼내기
                answer[j] = i - j; // 가격이 유지된 기간 계산
            }
            // 현재 인덱스를 스택에 추가
            stack.push(i);
        }

        // 스택에 남아있는 인덱스들은 끝까지 가격이 유지된 경우
        while (!stack.isEmpty()) {
            int j = stack.pop();
            answer[j] = n - 1 - j;
        }

        return answer;
    }

    public static void main(String[] args) {
        // ✅ 예제 테스트 실행
        System.out.println(Arrays.toString(solution(new int[]{1, 2, 3, 2, 3}))); // 결과: [4, 3, 1, 1, 0]
    }
}

File: 프로그래머스/2/42584. 주식가격/주식가격.js
function solution(prices) {
    // 가격 배열의 길이를 저장
    const n = prices.length;

    // 결과를 저장할 배열을 0으로 초기화
    const answer = new Array(n).fill(0);

    // 가격이 떨어지지 않은 구간의 시작 인덱스를 저장할 스택
    const stack = [];

    // 각 시간 인덱스를 하나씩 확인
    for (let i = 0; i < n; i++) {
        const v = prices[i];

        // 현재 가격이 이전 가격보다 낮은 경우,
        // 스택에 저장된 인덱스를 꺼내며 유지된 시간을 계산
        while (stack.length && prices[stack[stack.length - 1]] > v) {
            const j = stack.pop();     // 이전 시점의 인덱스를 꺼냄
            answer[j] = i - j;         // 현재 시간과의 차이를 저장
        }

        // 현재 인덱스를 스택에 추가
        stack.push(i);
    }

    // 스택에 남은 인덱스들은 끝까지 가격이 떨어지지 않은 경우
    while (stack.length) {
        const j = stack.pop();
        answer[j] = n - 1 - j;         // 전체 남은 시간을 계산하여 저장
    }

    // 결과 배열 반환
    return answer;
}

// ✅ 예제 테스트 실행
console.log(solution([1, 2, 3, 2, 3]));  // 결과: [4, 3, 1, 1, 0]
